### PR TITLE
Basic data parallel tests for ORTModule

### PR DESCRIPTION
**Pull request by wschin**
Continue of #6546.

### CODE REVIEW

File: orttraining/orttraining/test/python/orttraining_test_ort_module_pytorch_ddp.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    else:
+        print(f"  Rank {rank} uses Pytorch's nn.Module.");
+        model = ToyModel().to(rank)
+
+    ddp_model = DDP(model, device_ids=[rank])
+
+    loss_fn = nn.MSELoss()
+    optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)
+
+    CHECKPOINT_PATH = tempfile.gettempdir() + "/model.checkpoint"
```
Comment (wschin) [inline orttraining/orttraining/test/python/orttraining_test_ort_module_pytorch_ddp.py:101]:
```suggestion
    CHECKPOINT_PATH = os.path.join(tempfile.gettempdir(), "model.checkpoint")
```

File: orttraining/orttraining/test/python/orttraining_test_ort_module_pytorch_ddp.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
+# This test script is a modified version of Pytorch's tutorial.
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_ort_module_pytorch_ddp.py:1]: Would you please change the name of the file to follow template ```orttraining_test_ortmodule_test_nam```? Jus remove the ```_``` here.

Comment (wschin) [reply orttraining/orttraining/test/python/orttraining_test_ort_module_pytorch_ddp.py:1]: Np.

**Review by baijumeswani — DISMISSED** [review]

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_pytorch_ddp.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+def demo_basic(rank, world_size, use_ort_module):
+    torch.manual_seed(0)
+    print(f"Running basic DDP example on rank {rank}.")
+    setup(rank, world_size)
+
+    # create model and move it to GPU with id rank
+    if use_ort_module:
+        print(f"  Rank {rank} uses ORTModule.");
+        model = ToyModel().to(rank)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_pytorch_ddp.py:49]: this can be moved outside the if

Comment (wschin) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_pytorch_ddp.py:49]: What can be moved? to where?

**Review by baijumeswani — APPROVED** [review]
Looks good.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_pytorch_ddp.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    print(f"Rank {rank} sees loss {loss}")
+
+    if rank == 0:
+        assert torch.allclose(loss.cpu(), torch.FloatTensor([1.4909229278564453]))
+    elif rank == 1:
+        assert torch.allclose(loss.cpu(), torch.FloatTensor([1.0177688598632812]))
+    elif rank == 2:
+        assert torch.allclose(loss.cpu(), torch.FloatTensor([1.290669322013855]))
+    elif rank == 3:
+        assert torch.allclose(loss.cpu(), torch.FloatTensor([0.825118362903595]))
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_pytorch_ddp.py:132]: numeric tests tend to randomly fail on CI due to different versions of pytorch, cuda library or GPU hardware
We usually decreate tolerance to 1e-6 as opposed to the default 1e-7 to try preventing that

**Comment by thiagocrepaldi** [comment]
Is there a reason we should wrap ORTModule with DDP and not the other way around? Are both ways supported?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_pytorch_ddp.py (hunk 4 above)
Comment (wschin) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_pytorch_ddp.py:132]: This is a fairly small model and the result should be deterministic. Let's increase if we really see those random failures.

**Comment by wschin** [comment]
> Is there a reason we should wrap ORTModule with DDP and not the other way around? Are both ways supported?

Wrapping ORTModule under DDP is similar to wrapping ORTModule under Deepspeed. It's a proof that ORTModule works with existing distributed frameworks.

**Pull request merged**